lib/src/front_end/piece_factory.dart

```diff
@@ -853,14 +858,20 @@ mixin PieceFactory {
         }
       }
 
+      // Hoist any comments before the first operand so they don't force the
+      // infix operator to split.
+      if (operands.isEmpty) {
+        leadingComments = pieces.takeCommentsBefore(e.firstNonCommentToken);
+      }
+
       // Otherwise, just write the node itself.
       builder.visit(e);
     }
 
     traverse(node);
     operands.add(builder.build());
 
-    return InfixPiece(operands, indent: indent);
+    return InfixPiece(leadingComments, operands, indent: indent);
```

Comment:
How do we know that `leadingComments` has been definitely been assigned by this point?

Reply:
It's a little unclear since the code is recursive, but it's traversing a tree of binary expressions in order to flatten a chain of the same binary operator into a single list of operands (since we split them all together or not).

As it traverses, it will definitely hit leaf expressions at some point for all operands. And there will definitely be a first leaf expressions. So `leadingComments` will always get initialized. It's just hard for the compiler to see it because of the recursion.

But actually, now that I think about it, I don't need to wait until we've recursed into the left-most leaf operand to initialize `leadingComments`. The left-most operand's left-most token will always be the same as the left-most token of the entire `node`, so I can just initialize `leadingComments` from that. Fixed.